src/lib.rs

```diff
@@ -439,6 +442,64 @@ impl DstLayout {
         }
     }
 
+    /// Constructs a `DstLayout` which describes a dynamically-sized type with
+    /// the given properties.
+    ///
+    /// # Safety
+    ///
+    /// Unsafe code may assume that the returned `DstLayout` is the correct
+    /// layout for the described type so long as:
+    /// - The type has the alignment `align`
+    /// - The type has at least one field
+    /// - The byte offset of the first byte of the trailing field is equal to
+    ///   `trailing_field_offset`
+    /// - The trailing field's layout is correctly described by
+    ///   `trailing_field_layout`
+    #[doc(hidden)]
+    #[inline]
+    pub const fn for_dst(
+        align: NonZeroUsize,
+        trailing_field_offset: usize,
+        trailing_field_layout: DstLayout,
+    ) -> Option<DstLayout> {
+        Some(DstLayout {
+            // SAFETY: The caller has promised that this is the correct
+            // alignment.
+            _align: align,
+            _size_info: match trailing_field_layout._size_info {
+                SizeInfo::Sized { _size } => {
+                    let without_padding = match trailing_field_offset.checked_add(_size) {
+                        Some(without_padding) => without_padding,
+                        None => return None,
+                    };
+
+                    let padding = util::core_layout::_padding_needed_for(without_padding, align);
+
+                    let _size = match without_padding.checked_add(padding) {
+                        Some(_size) => _size,
+                        None => return None,
+                    };
+
+                    // SAFETY: TODO
+                    SizeInfo::Sized { _size }
```

Comment:
I'm not convinced this is sound for non-`repr(C)` types. Rust reserves the right to add excess padding.

Reply:
Oh yeah very good point. Maybe rename to `for_repr_c_dst` or something? And add a safety precondition?

Reply:
What should the UX of the derive be? Some possibilities:
- unconditionally require `repr(C)` on structs.
- don't unconditionally require `repr(C)`, but expand to `DstLayout::for_type::<Self>()` on non-`repr(C)` types, and `DstLayout::for_repr_c_dst` on `repr(C)` types.

Reply:
Generally I'd prefer the latter, but how can we make the UX of the error message good when the user tries to derive on an unsized, non-`repr(C)` type?